operator/internal/controller/charactercounter_controller.go

```diff
@@ -47,16 +55,138 @@ type CharacterCounterReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.0/pkg/reconcile
 func (r *CharacterCounterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	r.log = log.FromContext(ctx)
+	r.log.Info("start reconcile for", "name", req.Name)
 
-	// TODO(user): your logic here
+	var cc rampupv1alpha1.CharacterCounter
+	if err := r.Get(ctx, req.NamespacedName, &cc); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("unable to fetch ChracterCounter CRD: %w", err))
+	}
 
-	return ctrl.Result{}, nil
+	if cc.Spec.Namespace == "" {
+		cc.Spec.Namespace = "default"
+	}
+
+	r.cc = &cc
+
+	r.log.Info("server reconcileService")
+	if err := r.reconcileService(ctx, cc.Spec.Server); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("server reconcileDeployment")
+	if err := r.reconcileDeployement(ctx, cc.Spec.Server, false); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("frontend reconcileService")
+	if err := r.reconcileService(ctx, cc.Spec.Frontend); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("frontend reconcileDeployment")
+	return ctrl.Result{}, r.reconcileDeployement(ctx, cc.Spec.Frontend, true)
+}
+
+func (r *CharacterCounterReconciler) reconcileService(ctx context.Context, component rampupv1alpha1.CharacterCounterComponent) error {
+	var service corev1.Service
+	service.Name = component.Name
+	service.Namespace = r.cc.Namespace
+	service.Labels = r.cc.Spec.Labels
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, &service, func() error {
+		modifyService(*r.cc, component, &service)
+		return ctrl.SetControllerReference(r.cc, &service, r.Scheme)
+	})
+	if err != nil {
+		return fmt.Errorf("could not create or update service: %w", err)
+	}
+	if op != controllerutil.OperationResultNone {
+		r.log.Info("reconcile service successfully", "operation", op)
+	}
+	return nil
+}
+
+func modifyService(cc rampupv1alpha1.CharacterCounter, component rampupv1alpha1.CharacterCounterComponent, service *corev1.Service) {
+	service.Labels = cc.Spec.Labels
+	service.Spec = corev1.ServiceSpec{
+		Ports:    component.ServicePorts,
+		Selector: component.Selector,
+	}
+}
+
+func (r *CharacterCounterReconciler) reconcileDeployement(ctx context.Context, component rampupv1alpha1.CharacterCounterComponent, frontend bool) error {
+	args := []string{}
+	if frontend {
+		args = append(args, "-grpc-addr", fmt.Sprintf("%s:80", r.cc.Spec.Server.Name))
+	}
+	deployment := appsv1.Deployment{
```

Comment:
The first thing CreateOrUpdate does it `Get` which overwrites this deployment, so all of the definitions you add here are immediately gone either way if it exists or not.

Reply:
That wouldn't make much sense from how the docs are written.
The function only overwrites if the object does exist otherwise it returns non-empty error.
```
// CreateOrUpdate creates or updates the given object in the Kubernetes
// cluster. The object's desired state must be reconciled with the existing
// state inside the passed in callback MutateFn.
//
// The MutateFn is called regardless of creating or updating an object.
//
// It returns the executed operation and an error.
func CreateOrUpdate(ctx context.Context, c client.Client, obj client.Object, f MutateFn) (OperationResult, error) {
	key := client.ObjectKeyFromObject(obj)
	if err := c.Get(ctx, key, obj); err != nil {
		if !apierrors.IsNotFound(err) {
			return OperationResultNone, err
		}
		if err := mutate(f, key, obj); err != nil {
			return OperationResultNone, err
		}
		if err := c.Create(ctx, obj); err != nil {
			return OperationResultNone, err
		}
		return OperationResultCreated, nil

```

Reply:
Yes, but `Get` itself is mutating the object. It's better to let mutateFn handle filling in all necessary fields

Reply:
I just did a long search on github and debugging my code and could not find where the object is mutated when it does not exist already. Maybe you can point me in the right direction here :)

I just thought it is more efficient to reconcile only the smallest subset necessary, but if it is best practice to reconcile the entire object, perfekt! (I guess then I dont have to think about which fields are mutable or immutable.)

Reply:
No, you're right. I mixed things up. Get indeed only mutates when the object exists.

Premature optimization for a mutate function can lead to fields never being reset. Depends on how you want your ownership model to look like, but for most cases I would think that a controller that creates a resource should fully own it, so it should revert all unexpected modifications.